### PR TITLE
Fully azdevify the application

### PR DESCRIPTION
**Pull request by jongio**
This is a large PR, but was necessary because of all the additions/improvements that will come and need to be tested as a unit.

1. Adds infra/core bicep files
2. Adds Azure Developer CLI support
3. Optionally use KV

To effectively share code, we moved some projects into subfolders, so please see that.

Here are the tests:

With RBAC

AZURE_AUTH_TYPE= rbac
USE_KEY_VAULT= false
With Keys and not use Key Vault

AZURE_AUTH_TYPE= keys
USE_KEY_VAULT= false
With Keys and use Key Vault

AZURE_AUTH_TYPE= keys
USE_KEY_VAULT= true
Note: The current default values ​​are: AZURE_AUTH_TYPE=keys and USE_KEY_VAULT=true.

You can test with `azd init -t jongio/chat-with-your-data-solution-accelerator`

### CODE REVIEW

**Comment by ross-p-smith** [comment]
Thanks - really looking forward to this. When I run `azd up` it fails for me in the `web` package
image I will try and diagnose what that error is.

File: azure.yaml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+  template: chat-with-your-data-solution-accelerator@0.0.1-beta
+
+services:
+  web:
+    project: ./code
+    language: py
+    host: appservice
+    hooks:
+      prepackage:
+        windows:
```
Comment (ross-p-smith) [inline azure.yaml:14]: I believe you need to add something like: 
```yaml
        posix:
          shell: sh
          run:  cd ./frontend;npm install;npm run build;
          interactive: true
          continueOnError: false
```

File: infra/main.bicep [hunk 2]
```diff
@@ -0,0 +1,746 @@
...
+param azureCognitiveSearchSku string = 'standard'
+
+@description('Azure Cognitive Search Index')
+param azureSearchIndex string = '${environmentName}-index-${resourceToken}'
+
+@description('Azure Cognitive Search Conversation Log Index')
+param azureSearchConversationLogIndex string = 'conversations'
+
+@description('Name of Storage Account')
+param storageAccountName string = '${environmentName}str'
```
Comment (ross-p-smith) [inline infra/main.bicep:134]: Shouldn't this have a ${resourceToken} otherwise it resolves to `devstr` which won't be a unique storage account name.

**Review by ross-p-smith — CHANGES_REQUESTED** [review]

File: azure.yaml (hunk 1 above)
Comment (jongio) [reply azure.yaml:14]: thanks, are you running in devcontainer?

Comment (ross-p-smith) [reply azure.yaml:14]: yes

File: infra/main.bicep [hunk 3]
```diff
@@ -0,0 +1,746 @@
...
+    speechKeyName: useKeyVault ? storekeys.outputs.SPEECH_KEY_NAME: ''
+    useKeyVault: useKeyVault
+    keyVaultName: useKeyVault || authType == 'rbac' ? keyvault.outputs.name : ''
+    keyVaultEndpoint: useKeyVault ? keyvault.outputs.endpoint : ''
+    authType: authType
+    appSettings: {
+      AZURE_SEARCH_SERVICE: 'https://${azureCognitiveSearchName}.search.windows.net'
+      AZURE_SEARCH_INDEX: azureSearchIndex
+      AZURE_SEARCH_CONVERSATIONS_LOG_INDEX: azureSearchConversationLogIndex
+      AZURE_SEARCH_SEMANTIC_SEARCH_CONFIG: azureSearchSemanticSearchConfig
```
Comment (ross-p-smith) [inline infra/main.bicep:341]: I believe you are missing `AZURE_SEARCH_USE_SEMANTIC_SEARCH` value. It might be worth going through the changes that were made to deployment.bicep since you forked this as main.bicep seems to be a clean sheet of paper (I would have done the same! :-) )

File: infra/main.bicep (hunk 2 above)
Comment (ross-p-smith) [reply infra/main.bicep:134]: I changed this to
`${environmentName}str${resourceToken}` and my worked (but my environment is simply dev)

File: infra/main.bicep [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+@description('Whether or not to stream responses from Azure OpenAI')
+param azureOpenAIStream string = 'true'
+
+@description('Azure OpenAI Embedding Model Deployment Name')
+param azureOpenAIEmbeddingModel string = 'text-embedding-ada-002'
+
+@description('Azure OpenAI Embedding Model Name')
+param azureOpenAIEmbeddingModelName string = 'text-embedding-ada-002'
+
+@description('Azure Cognitive Search Resource')
```
Comment (ross-p-smith) [inline infra/main.bicep:114]: Please remove all references to `Azure Cognitive Search` as the product is now `Azure AI Search`. (I haven't listed all occurrences in this file)

**Comment by ross-p-smith** [comment]
The files in `code/static` should be deleted and the .gitignore should be updated to read 
(Line 7)

```text
# generated frontend files
code/static/
```

File: code/app.py [hunk 5]
```diff
@@ -26,29 +30,39 @@
 def static_file(path):
     return app.send_static_file(path)
 
-
-@app.route("/api/config", methods=["GET"])
+AZURE_AUTH_TYPE = os.environ.get("AZURE_AUTH_TYPE", "keys")
```
Comment (ross-p-smith) [inline code/app.py:33]: The code base has had `black` and `flake8` run against it. If you do this, then we may see less merge conflicts

File: code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py [hunk 6]
```diff
@@ -1,12 +1,9 @@
 from typing import Optional
 from datetime import datetime, timedelta
-from azure.storage.blob import (
```
Comment (ross-p-smith) [inline code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py:3]: Most of the tests that are now in the root will have to have their import changed to reflect this folder move. This is also the reason for the failing build. (Adding `.backend.batch` should fix it!)

**Comment by ross-p-smith** [comment]
I can successfully deploy the solution (great improvement by the way, azd is definitely the way forward). I can also upload a single file using the Admin App (tick) and I can see it in Storage. I can't however see the data or delete the data from the UI, so I feel that there is a permission problem there. I'm using keys and I can see the secrets in keyvault and they look correct. I can also see a message in the Storage Queue, but the Function says it is running, but has not taken the message from the Queue.

File: .env.sample [hunk 7]
```diff
@@ -49,4 +49,7 @@ ORCHESTRATION_STRATEGY=openai_functions
 #Speech-to-text feature
 AZURE_SPEECH_SERVICE_KEY=
 AZURE_SPEECH_SERVICE_REGION=
-AZURE_AUTH_TYPE=rbac
\ No newline at end of file
+# Auth type related param. When your AZURE_AUTH_TYPE = rbac, please make sure setting USE_KEY_VAULT= false; When your USE_KEY_VAULT= true, please setting your key vault endpoint.
```
Comment (pamelafox) [inline .env.sample:52]:
```suggestion
# Auth type environment variables.
# When AZURE_AUTH_TYPE=rbac, please make sure variable USE_KEY_VAULT=false
# When USE_KEY_VAULT = true, please make sure to set AZURE_KEY_VAULT_ENDPOINT
```

Rewording for clarity.


FYI, for azure-search-openai-demo, we sometimes have situations where we use RBAC for the majority but keyvault for a few SDKs, due to lack of RBAC support for that SDK or SKU. So you may imagine a world where that happens for this SDK, and the key vault endpoint isn't mutually exclusive. Fine for now though if you managed to make a clean separation!

File: README.md [hunk 8]
```diff
@@ -194,6 +194,49 @@ az deployment group create --resource-group $RESOURCE_GROUP_NAME --template-file
...
+powershell -ex AllSigned -c "Invoke-RestMethod 'https://aka.ms/install-azd.ps1' | Invoke-Expression"
+```
+
+### Linux/MacOS
+
+```
+curl -fsSL https://aka.ms/install-azd.sh | bash
+```
+
+After logging in with the following command, you will be able to use the `azd` cli to quickly provision and deploy the application.
```
[the remark this answers is not in the document]
Comment (pamelafox) [reply README.md:215]: Agreed! They will need to use the feature to get it installing properly in the dev container on an M1, so they shouldn't re-run that if they're in a dev container.

File: README.md [hunk 9]
```diff
@@ -194,6 +194,49 @@ az deployment group create --resource-group $RESOURCE_GROUP_NAME --template-file
...
+```
+azd env set AUTH_TYPE rbac
+```
+
+Then, run `azd up` to provision all the resources to Azure and deploy the code to those resources.
+```
+azd up 
+```
+
+Select your desired `subscription` and `location`. Then choose a resource group or create a new resource group. Wait a moment for the resource deployment to complete, click the Website endpoint and you will see the web app page.
```
Comment (pamelafox) [inline README.md:237]:
```suggestion
Select your desired `subscription` and `location`. Then choose a resource group or create a new resource group. Wait a moment for the resource deployment to complete, click the website endpoint and you will see the web app page.
```

It prompts you to choose a resource group? (I haven't checked infra yet, just surprised).

File: code/app.py [hunk 10]
```diff
@@ -2,10 +2,14 @@
 import os
 import logging
 import requests
-import openai
+from openai import AzureOpenAI
+
+# Fixing MIME types for static files under Windows
```
Comment (pamelafox) [inline code/app.py:7]: Accidental repeated comment from below

File: code/app.py [hunk 11]
```diff
@@ -26,29 +30,39 @@
...
+AZURE_AUTH_TYPE = os.environ.get("AZURE_AUTH_TYPE", "keys")
+# Initialize Azure keys based on authentication type and environment settings. 
+# When AZURE_AUTH_TYPE is not 'rbac' and USE_KEY_VAULT environment variable is set, 
+# Azure keys are securely fetched from Azure Key Vault using DefaultAzureCredential. 
+# Otherwise, keys are obtained from environment variables, with fallbacks to None or 
+# an empty string depending on the AZURE_AUTH_TYPE.
+if not AZURE_AUTH_TYPE == 'rbac' and os.environ.get("USE_KEY_VAULT"):
+    credential = DefaultAzureCredential()
+    secret_client = SecretClient(os.environ.get("AZURE_KEY_VAULT_ENDPOINT"), credential)
+    AZURE_SEARCH_KEY = secret_client.get_secret(os.environ.get("AZURE_SEARCH_KEY")).value
```
Comment (pamelafox) [inline code/app.py:42]: You might want to first make sure the environment variable exists and then try to fetch it, so that you don't try to fetch a secret called None. It's an error either way but catching earlier is easier debugging.

File: code/app.py [hunk 12]
```diff
@@ -26,29 +30,39 @@
...
+    AZURE_OPENAI_KEY = secret_client.get_secret(os.environ.get("AZURE_OPENAI_KEY")).value
+    AZURE_SPEECH_KEY = secret_client.get_secret(os.environ.get("AZURE_SPEECH_SERVICE_KEY")).value
+else:
```
Comment (pamelafox) [inline code/app.py:45]: We generally recommend avoiding Python ternaries for readability. I think it'd be clearer to do

elif AZURE_AUTH_TYPE == 'rbac':
  AZURE_SEARCH_KEY = None
  AZURE_OPENAI_KEY = ""
  AZURE_SPEECH_KEY = None
else:
  AZURE_SEARCH_KEY = os.environ.get("AZURE_SEARCH_KEY")
  AZURE_OPENAI_KEY = os.environ.get("AZURE_OPENAI_KEY")
  AZURE_SPEECH_KEY = os.environ.get("AZURE_SPEECH_SERVICE_KEY")

I'd probably put the rbac case first and then the two key cases.

Just a style thing though, not a big deal. Python ternaries are unfortunately not the most readable, the way they're ordered.

File: code/app.py [hunk 13]
```diff
@@ -26,29 +30,39 @@
...
+    azure_speech_region = os.getenv('AZURE_SPEECH_SERVICE_REGION')
 
     # Return the configuration data as JSON
-    return jsonify(
-        {"azureSpeechKey": azure_speech_key, "azureSpeechRegion": azure_speech_region}
-    )
-
+    return jsonify({
+        'azureSpeechKey': AZURE_SPEECH_KEY,
+        'azureSpeechRegion': azure_speech_region
```
Comment (pamelafox) [inline code/app.py:58]: Inconsistent use of single and double quotes in this file.

File: code/app.py (hunk 11 above)
Comment (ross-p-smith) [reply code/app.py:42]: Could we push this down to EnvHelper as this code is repeated there too

File: code/app.py [hunk 14]
```diff
@@ -241,20 +252,21 @@ def stream_without_data(response):
...
 
     for message in request_messages:
         messages.append({"role": message["role"], "content": message["content"]})
 
-    response = openai.ChatCompletion.create(
-        engine=AZURE_OPENAI_MODEL,
-        messages=messages,
+    response = openai_client.chat.completions.create(
+        model=AZURE_OPENAI_MODEL,
+        messages = messages,
```
Comment (pamelafox) [inline code/app.py:269]: Inconsistent spacing around = signs

```suggestion
        messages=messages,
```

File: code/app.py [hunk 15]
```diff
@@ -241,20 +252,21 @@ def stream_without_data(response):
...
     messages = [{"role": "system", "content": AZURE_OPENAI_SYSTEM_MESSAGE}]
 
     for message in request_messages:
         messages.append({"role": message["role"], "content": message["content"]})
 
-    response = openai.ChatCompletion.create(
-        engine=AZURE_OPENAI_MODEL,
-        messages=messages,
+    response = openai_client.chat.completions.create(
+        model=AZURE_OPENAI_MODEL,
```
Comment (pamelafox) [inline code/app.py:268]: We usually make a note in our ChatCompletion calls that Azure Open AI takes the deployment name as the model name. I guess in the repo "AZURE_OPENAI_MODEL" means deployment name. It might help to clarify that somewhere, as MODEL and MODEL_NAME sound awfully similar.

File: code/app.py [hunk 16]
```diff
@@ -331,20 +342,9 @@ def conversation_custom():
...
-                )
-        from utilities.helpers.ConfigHelper import ConfigHelper
-
-        messages = message_orchestrator.handle_message(
-            user_message=user_message,
-            chat_history=chat_history,
-            conversation_id=conversation_id,
-            orchestrator=ConfigHelper.get_active_config_or_default().orchestrator,
-        )
+                chat_history.append((user_assistant_messages[i]['content'],user_assistant_messages[i+1]['content']))
```
Comment (pamelafox) [inline code/app.py:345]:
```suggestion
                chat_history.append((user_assistant_messages[i]["content"], user_assistant_messages[i+1]["content"]))
```

I'm sorry, I get so distracted by style inconsistencies. I think this repo doesn't enforce style linting yet, but perhaps you can manually lint your additions to match the existing style in terms of spaces and quotes.
Or just ignore my style nits.

File: code/app.py [hunk 17]
```diff
@@ -331,20 +342,9 @@ def conversation_custom():
...
+        from backend.batch.utilities.helpers.ConfigHelper import ConfigHelper
+        messages = message_orchestrator.handle_message(user_message=user_message, chat_history=chat_history, conversation_id=conversation_id, orchestrator=ConfigHelper.get_active_config_or_default().orchestrator)
```
Comment (pamelafox) [inline code/app.py:347]: Another line length change? (I think it's nice to avoid those in diffs unless it's due to an explicit style change, as it makes it harder to see what actually changed).

File: code/app.py [hunk 18]
```diff
@@ -370,4 +370,4 @@ def conversation_custom():
 
 
 if __name__ == "__main__":
-    app.run()
+    app.run()
```
Comment (pamelafox) [inline code/app.py:373]: Missing newline at end of file

File: code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py [hunk 19]
```diff
@@ -131,9 +121,10 @@ def get_all_files(self):
         return files
 
     def upsert_blob_metadata(self, file_name, metadata):
-        blob_client = BlobServiceClient.from_connection_string(
-            self.connect_str
-        ).get_blob_client(container=self.container_name, blob=file_name)
+        if self.auth_type == 'rbac':
+            blob_client = self.blob_service_client.get_blob_client(container=self.container_name, blob=file_name)
```
Comment (pamelafox) [inline code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py:125]: Just curious, why not make the blob service client in the constructor? For timing reasons?

File: code/app.py (hunk 16 above)
Comment (ross-p-smith) [reply code/app.py:345]: black and flake8 have recently been introduced as a git-precommit. Looks like it has not been installed (or not running from the devcontainer). I think this will greatly help a lot of the style comments you are mentioning

Comment (pamelafox) [reply code/app.py:345]: Ah ross just said that the codebase has black now, yay!

File: code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py (hunk 6 above)
Comment (zedy-wj) [reply code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py:3]: Thanks! I ignored here. :)

**Comment by zedy-wj** [comment]
It looks like a hierarchy change in the `code` folder is causing the current CI error.
image

File: code/app.py (hunk 11 above)
Comment (zedy-wj) [reply code/app.py:40]: In fact, in the [bicep file](https://github.com/jongio/chat-with-your-data-solution-accelerator/blob/main/infra/app/adminweb.bicep#L38-L44), all `AZURE_XXX_KEY` are associated with `USE_KEY_VAULT`, so `get_secret(None)` will not appear here.

File: code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py (hunk 19 above)
Comment (zedy-wj) [reply code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py:125]: `blob_service_client` is already in the constructor, what do you want to say is `blob_client`? This is the author's initial design, and we have not changed it.

File: code/app.py (hunk 17 above)
Comment (zedy-wj) [reply code/app.py]: Yes, I ignored these formatting changes during the last merge conflict and have now removed all line length related changes. Sorry for the inconvenience.

File: infra/main.bicep (hunk 2 above)
Comment (zedy-wj) [reply infra/main.bicep]: Modified according to @ross-p-smith suggestion, it is worth mentioning that `${resourceToken}` is usually a random string of 13 characters in length and the storage name has a maximum limit of 24 characters. This means that the `${environmentName}` need to be simplified a bit. :)

File: code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py (hunk 19 above)
Comment (pamelafox) [reply code/backend/batch/utilities/helpers/AzureBlobStorageHelper.py:153]: Oh sorry thats what I meant. Okay, that's fine, was just curious.

**Comment by ross-p-smith** [comment]
Thanks for the update @zedy-wj ! I've just pushed a small change which will get the tests passing. Within the devcontainer we install all of the python requirements when we create the container (look at `postCreate.sh`) - this still had a reference to a file that you have removed and it also showed that some of the dependencies had been updated since you forked and were then conflicting with the Teams extension which I don't believe you have touched.

As far as your question about the posix code snippet - yes I can deploy the Azure resources from a devcontainer with that !

I'll start reviewing again  . . .

**Comment by jongio** [comment]
> As far as your question about the posix code snippet - yes I can deploy the Azure resources from a devcontainer with that !

What are you referring to here?

**Comment by ross-p-smith** [comment]
> What are you referring to here?

I was answering zedy-wj's question about that I can now deploy from a devcontainer since the update.

**Comment by zedy-wj** [comment]
@ross-p-smith - The function-related issues have been fixed, could you try again?

@jongio - CI errors and comments have been resolved so far.

File: .vscode/launch.json [hunk 20]
```diff
@@ -17,7 +17,7 @@
                 "run",
                 "Admin.py"
             ],
-            "cwd": "${workspaceFolder}/code/admin",
```
Comment (ross-p-smith) [inline .vscode/launch.json:20]: When we are using azd an .env file is generated under the .azure folder - do we have to add something like
```
"envFile": "${workspaceFolder}/.azure/dev/.env"
```
to each of these launch - or is there an azd way of doing this?

File: code/app.py [hunk 21]
```diff
@@ -241,19 +246,21 @@ def stream_without_data(response):
 
 
 def conversation_without_data(request):
-    openai.api_type = "azure"
-    openai.api_base = f"https://{AZURE_OPENAI_RESOURCE}.openai.azure.com/"
-    openai.api_version = "2023-03-15-preview"
-    openai.api_key = AZURE_OPENAI_KEY
+    azure_endpoint = f"https://{AZURE_OPENAI_RESOURCE}.openai.azure.com/"
+    if AZURE_AUTH_TYPE == "rbac":
+        openai_client = AzureOpenAI(azure_endpoint=azure_endpoint,  api_version=AZURE_OPENAI_API_VERSION, azure_ad_token_provider=AZURE_TOKEN_PROVIDER)
```
Comment (ross-p-smith) [inline code/app.py:251]: Extra space here. I think you need to run `black code` from a terminal. Or run `pre-commit install`. I ran this and there are 10 files that need formatting

File: code/backend/batch/requirements.txt [hunk 22]
```diff
@@ -1,18 +1,28 @@
...
 fake-useragent==1.4.0
 chardet==5.2.0
 --extra-index-url https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-python/pypi/simple/
 azure-search-documents==11.4.0b8
 opencensus-ext-azure==1.1.13
-azure-ai-contentsafety==1.0.0
+pandas==1.5.1
 python-docx==1.1.0
+azure-keyvault-secrets==4.4.*
+# Add dev dependencies here - this will be refactored out by Poetry
```
Comment (ross-p-smith) [inline code/backend/batch/requirements.txt:27]: pytest has been moved to dev-requirements.txt - can you check all the versions as dependabot has been turned on and some of these versions have been updated. e.g. scipy, scikit-learn, transformers, etc

**Review by ross-p-smith — CHANGES_REQUESTED** [review]

File: infra/core/monitor/loganalytics.bicep [hunk 23]
```diff
@@ -0,0 +1,22 @@
+metadata description = 'Creates a Log Analytics workspace.'
+param name string
+param location string = resourceGroup().location
+param tags object = {}
+
+resource logAnalytics 'Microsoft.OperationalInsights/workspaces@2021-12-01-preview' = {
```
Comment (ross-p-smith) [inline infra/core/monitor/loganalytics.bicep:6]: should this be preview?

File: infra/core/security/role.bicep [hunk 24]
```diff
@@ -0,0 +1,21 @@
+metadata description = 'Creates a role assignment for a service principal.'
```
Comment (ross-p-smith) [inline infra/core/security/role.bicep:1]: Can we remove the one in the directory above? And are we going to remove the old deployment.bicep as part of this PR? Or if you want to make it a part of another PR can you create an issue and link it here so we dont forget?

**Comment by ross-p-smith** [comment]
I've tested this and I'm happy with it (after the requested changes are completed). 

@hunterjam are you able to test the Teams part?

@adamdougal - are you able to get this across the line whilst I'm away?

File: code/app.py (hunk 21 above)
Comment (ross-p-smith) [reply code/app.py]: I've run black and then pushed!

File: infra/core/monitor/loganalytics.bicep (hunk 23 above)
Comment (zedy-wj) [reply infra/core/monitor/loganalytics.bicep:6]: @jongio - I checked the files in the latest core and it is still this version. Do we need to change it?

File: infra/core/security/role.bicep (hunk 24 above)
Comment (zedy-wj) [reply infra/core/security/role.bicep:1]: > Can we remove the one in the directory above?

Are you referring to the role.bicep file? We need to use it when configuring rbac permissions.


> And are we going to remove the old deployment.bicep as part of this PR? Or if you want to make it a part of another PR can you create an issue and link it here so we dont forget?

Created an issue and linked here.

File: .vscode/launch.json (hunk 20 above)
Comment (zedy-wj) [reply .vscode/launch.json:20]: Usually we don't do this. The parameters in the `.env` file generated under the `.azure` folder are not enough to deploy the application. For security reasons, all keys will not appear in this file. (Except for one case: when you use `Key Vault`, all Keys at this time are just KeyName, so we can output them)

File: code/backend/batch/requirements.txt (hunk 22 above)
Comment (zedy-wj) [reply code/backend/batch/requirements.txt:27]: Updated.

**Comment by zedy-wj** [comment]
@ross-p-smith - Updated according to your comments. Please re-review it, thanks!

**Review by adamdougal — DISMISSED** [review]
Good job! I can appreciate the effort that has gone into this ❤️ 

I've added some inline comments, but code wise it generally looks good. Unfortunately I was unable to fully test due to an issue running the admin app. 

Maybe not as part of this PR, but I think it would be good in future to add some documentation around using KeyVault. But also maybe to refactor some of these changes to use the strategy pattern rather than the if/else blocks that are dotted around.

**Comment by adamdougal** [comment]
> This problem should be solved if we place the tests in the code folder. What do you think?

I think that's okay. 

As an alternative, though it might break the deployment. If you change everywhere in `app.py` that imports from the backend dir to be prepended with `code` it fixes the issue. E.g.

Change this:
```python
from backend.batch.utilities.helpers.EnvHelper import EnvHelper
```

To this:
```python
from code.backend.batch.utilities.helpers.EnvHelper import EnvHelper
```

I believe this is because of line 17 of the `app.py`:
```python
sys.path.append(os.path.join(os.path.dirname(__file__), ".."))
```

This fixes the tests, however I've not tested the deployment with this change.

File: infra/main.bicep [hunk 25]
```diff
@@ -0,0 +1,751 @@
...
+param azureSearchSku string = 'standard'
+
+@description('Azure AI Search Index')
+param azureSearchIndex string = '${environmentName}-index-${resourceToken}'
+
+@description('Azure AI Search Conversation Log Index')
+param azureSearchConversationLogIndex string = 'conversations'
+
+@description('Name of Storage Account')
+param storageAccountName string = '${environmentName}str${resourceToken}'
```
[the remark this answers is not in the document]
Comment (jongio) [reply infra/main.bicep:134]: @zedy-wj - Let's do what we do in other azd templates where we don't have issues.  Thanks.  I don't think we use env name in storage account name because that can be very long.

File: infra/core/monitor/loganalytics.bicep (hunk 23 above)
Comment (jongio) [reply infra/core/monitor/loganalytics.bicep:6]: Yes, there are features in that preview that are needed.  We can dig in more if needed.

File: README.md [hunk 26]
```diff
@@ -194,6 +194,46 @@ az deployment group create --resource-group $RESOURCE_GROUP_NAME --template-file
...
+Enter an environment name.
+
+**Notes:** the default auth type uses keys, if you want to switch to rbac, please run `azd env set AUTH_TYPE rbac`.
+
+Then, run `azd up` to provision all the resources to Azure and deploy the code to those resources.
+```
+azd up 
+```
+
+Select your desired `subscription` and `location`. Wait a moment for the resource deployment to complete, click the website endpoint and you will see the web app page.
```
[the remark this answers is not in the document]
Comment (zedy-wj) [reply README.md:234]: @jongio - Three services are currently used. In local deployment, we also noticed that packaging in `python` and `npm` takes a lot of time. In fact, `adminweb` and `function` services use the same `requirements.txt` file, but when using azd to deploy, because two independent services need to install packages twice, the deployment time is too long. Do you have any suggestions for optimizing deployment time?

File: docs/LOCAL_DEPLOYMENT.md [hunk 27]
```diff
@@ -91,14 +91,13 @@ This step is included if you cannot use the Launch configuration in VSCode. Open
 ```shell
 cd code
 python -m pip install -r requirements.txt
-cd app
 python -m flask --app ./app.py --debug run
```
[the remark this answers is not in the document]
Comment (zedy-wj) [reply docs/LOCAL_DEPLOYMENT.md:94]: We also encountered this problem during local deployment. After preliminary investigation, it may be caused by changes in the directory structure. You just need to use `pip install InteractiveInterpreter`, reinstall this package. Of course, we will continue to follow up on this issue.

File: code/backend/batch/requirements.txt [hunk 28]
```diff
@@ -0,0 +1,28 @@
+azure-functions==1.18.0
```
[the remark this answers is not in the document]
Comment (zedy-wj) [reply code/backend/batch/requirements.txt:1]: For azd deploy, `adminweb` and `function` use two independent services, which also means that they require independent `requirements.txt` files. I understand that these two are duplicate files. If you want to use only one, we can put `batch` and `adminweb` at the same level as in the previous directory structure. Do you think this is okay?
Such as:
image

File: infra/app/adminweb.bicep [hunk 29]
```diff
@@ -0,0 +1,58 @@
...
+param openAIKeyName string = ''
+param storageAccountKeyName string = ''
+param formRecognizerKeyName string = ''
+param searchKeyName string = ''
+param contentSafetyKeyName string = ''
+param speechKeyName string = ''
+param keyVaultEndpoint string = ''
+param authType string
+
+module adminweb '../core/host/appservice.bicep' = {
```
[the remark this answers is not in the document]
Comment (zedy-wj) [reply infra/app/adminweb.bicep:28]: Sorry, I can't reproduce this problem. I guess you may have encountered network problems that caused the failure during the packaging process, because I checked that the [streamlit module](https://github.com/jongio/chat-with-your-data-solution-accelerator/blob/main/code/backend/requirements.txt#L2) is indeed included in the package.

File: README.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply README.md:234]: Could you have a step that installs the requirements first and then uses a multi-stage docker build step that shares the locations of the packages?

The devcontainer installs all of these requirements as a PostCreate step so as long as we use a convention we can fix it for this too

File: infra/app/adminweb.bicep (hunk 29 above)
[the remark this answers is not in the document]
Comment (adamdougal) [reply infra/app/adminweb.bicep:28]: Let me try this again from a clean dev container

File: code/backend/pages/01_Ingest_Data.py [hunk 30]
```diff
@@ -88,39 +110,79 @@ def upload_file(bytes_data: bytes, file_name: str, content_type: Optional[str] =
         )
         content_type = content_type if content_type is not None else "text/plain"
     account_name = os.getenv("AZURE_BLOB_ACCOUNT_NAME")
-    account_key = os.getenv("AZURE_BLOB_ACCOUNT_KEY")
-    container_name = os.getenv("AZURE_BLOB_CONTAINER_NAME")
-    if account_name is None or account_key is None or container_name is None:
-        raise ValueError(
-            "Please provide values for AZURE_BLOB_ACCOUNT_NAME, AZURE_BLOB_ACCOUNT_KEY and AZURE_BLOB_CONTAINER_NAME"
+    if os.environ.get("AUTH_TYPE") == "rbac":
+        credential = DefaultAzureCredential()
```
[the remark this answers is not in the document]
Comment (zedy-wj) [reply code/backend/pages/01_Ingest_Data.py]: We're looking for ways to optimize this. :)

File: docs/LOCAL_DEPLOYMENT.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (zedy-wj) [reply docs/LOCAL_DEPLOYMENT.md:94]: This bug is caused by changes in the directory structure. Since the `app.py` file is currently in the root directory of the `code` folder, and the `werkzeug` package import the `code` in the system lib, it is considered to have been rewritten if it has the same name. I changed `code` folder to `codes` and this problem no longer exists.

**Comment by zedy-wj** [comment]
Sorry for the delay.

> To this:
> 
> ```python
> from code.backend.batch.utilities.helpers.EnvHelper import EnvHelper
> ```
I tested the approach like this and for tests it worked for me, but unfortunately `azd` is still not able to deploy.


@ross-p-smith ,@adamdougal - At present, I moved the `tests` folder to the `codes` folder. Following the steps as below, I can run the test successfully in local, but it fails in CI. Is it a problem with the root path?
image

File: .vscode/launch.json (hunk 20 above)
Comment (ross-p-smith) [reply .vscode/launch.json:20]: Will pick this up in #69

File: code/backend/batch/requirements.txt (hunk 28 above)
[the remark this answers is not in the document]
Comment (ross-p-smith) [reply code/backend/batch/requirements.txt]: I think we get this PR in and then address this as part of #170

**Comment by ross-p-smith** [comment]
> At present, I moved the `tests` folder to the `codes` folder. 

I'm not sure we want to rename the `code` folder to `codes`. If we need to move the `tests` folder then this is OK, as long as it is not alongside the code that it is testing so that it does not get copied into the docker images.

File: infra/core/storage/storage-account.bicep [hunk 31]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  resource queueServices 'queueServices' = if (!empty(queues)) {
+    name: 'default'
+    properties: {
+      cors: {
+        corsRules: []
+      }
+    }
+    resource queue 'queues' = [for queue in queues: {
```
Comment (ross-p-smith) [inline infra/core/storage/storage-account.bicep:70]: If you deploy this a second time you get a problem with the Event Grid "System topic source cannot be modified." - I think that this is the same as #280

Comment (zedy-wj) [reply infra/core/storage/storage-account.bicep:70]: Let's track this issue in #280.

File: README.md (hunk 26 above)
[the remark this answers is not in the document]
Comment (zedy-wj) [reply README.md:234]: > Could you have a step that installs the requirements first and then uses a multi-stage docker build step that shares the locations of the packages?

@ross-p-smith - If I understand correctly, you mean to install the required `requirements` in a container first, and then use docker to obtain them in other containers where they are needed, right?

@jongio - Do you have any ideas about the above requirements? As far as I know, when using `azd deploy`, the `host=appservice` with `language=py` will first find the requirements in the current path and install it. And using `azd`, the process from provisioning resources to deploying resources is consistent. If we want to pull other images in appservice, it seems that we can only do some operations through the `appCommandLine` parameter during provisioning (docker has been installed in appservice). What do you think about this? Or do you have better optimization suggestions?

File: docs/LOCAL_DEPLOYMENT.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (zedy-wj) [reply docs/LOCAL_DEPLOYMENT.md:94]: The above method has been abandoned. We have solved it in other ways. You can retest it, thanks!

**Comment by zedy-wj** [comment]
> I'm not sure we want to rename the `code` folder to `codes`.

Sorry, the folder name has been changed back. Currently, through the following modifications, Unit Tests can pass normally. The error in `CI/build` seems to be caused by not switching into the `code` folder.
image

I only see the `imageName`, @ross-p-smith do you know where to update it?
image

**Review by ross-p-smith — APPROVED** [review]
Thank you ever so much for this - I understand that this took great effort
